Fixes #7766: ms dhcp provider uses native dhcps api now.

dmitri-d opened the pull request.

dmitri-d committed:
Fixes #7766: ms dhcp provider uses native dhcps api now.

dmitri-d commented:
@pikelly, @alexjfisher, @timogoebel: This is a rewrite of ms dhcp provider to use native dhcps api. Could you give it a test when you have a chance please?

Note: I suggested to remove code related to vendor options from ms dhcp provider on the mailing list a couple of weeks ago, this PR removes this functionality.

timogoebel commented:
@witlessbird : I'd love to help, but I don't have a ms dhcp server at hand, sorry.

dmitri-d commented:
@timogoebel: oh sorry, for some reason I thought you use ms stuff.

pikelly commented:
This poor timing! I am on vacation now for a couple of weeks. :-(

alexjfisher commented:
Me too! ;)

Sent from my iPhone

On 4 Aug 2016, at 19:17, Paul Kelly notifications@github.com wrote:

> This poor timing! I am on vacation now for a couple of weeks. :-(
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

pikelly commented:
Any advice? I am digging in myself but this is Windows FFI. Heady stuff!!

C:\Proxy\brs>irb
irb(main):001:0> require 'dhcpsapi'
FFI::NotFoundError: Function 'DhcpV4EnumSubnetReservations' not found in [dhcpsapi]
        from C:/Ruby200/lib/ruby/gems/2.0.0/gems/ffi-1.9.10-x86-mingw32/lib/ffi/library.rb:261:in `attach_function'
        from C:/Ruby200/lib/ruby/gems/2.0.0/gems/dhcpsapi-0.0.3/lib/dhcpsapi/reservation.rb:60:in`module:DhcpsApi'
        from C:/Ruby200/lib/ruby/gems/2.0.0/gems/dhcpsapi-0.0.3/lib/dhcpsapi/reservation.rb:1:in `<top (required)>'
        from C:/Ruby200/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in`require'
        from C:/Ruby200/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from C:/Ruby200/lib/ruby/gems/2.0.0/gems/dhcpsapi-0.0.3/lib/dhcpsapi.rb:13:in`<top (required)>'
        from C:/Ruby200/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:135:in `require'
        from C:/Ruby200/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:135:in`rescue in require'
        from C:/Ruby200/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:144:in `require'
        from (irb):1
        from C:/Ruby200/bin/irb:12:in`<main>'
irb(main):002:0> exit

C:\Proxy\brs>bundle list
Gems included by the bundle:
- bundler (1.12.5)
- dhcpsapi (0.0.3)
- ffi (1.9.10)
- highline (1.7.8)
- json (1.8.3)
- rack (1.5.5)
- rack-protection (1.5.3)
- rb-inotify (0.9.7)
- rb-kqueue (0.2.4)
- sinatra (1.4.7)
- smart_proxy (1.13.0.develop)
- tilt (2.0.5)
- win32-service (0.8.7)

C:\Proxy\brs>ruby --version
ruby 2.0.0p648 (2015-12-16) [i386-mingw32]

dmitri-d commented:
@pikelly: What windows version are you on? This call requires win2012 server...

pikelly commented:
Unfortunately my test system is W2K8R2. There is no change of getting one of these today :-(

dmitri-d commented:
Hmm. There are a few more calls that I use that are available starting with windows server 2012. Considering that unless one has extended support contract, windows server 2008 reached end of support (or about to, depending on the version) -- does this extra effort make sense (would probably take me a couple of days)?

pikelly commented:
The problem with large organizations with many, many of these smart proxies scattered around the world is that it takes a really long time to get everyone to replace their existing hardware. This change does not allow for a drop in replacement but I am grateful for whatever is released.

Do you have any figures on performance improvement as this is our main concern with the MS DHCP proxies. Another one is the requirement for the proxy to be on the same LAN as the DHCP server, again for performance reasons.

dmitri-d commented:
@pikelly: I'll check how many calls I'd need to add and let you know. Re: performance -- I'm testing against a windows 2012 server running in a vm. The server only has a couple of networks with a few hosts in each. The calls are faster, but I'm unable to collect any performance-related stats from such a setup. I expect the performance to be on par with that of MS dhcp manager (as i'm using the same calls).

dmitri-d commented:
@pikelly: Looks like there is only two Windows 2012-specific calls that I use. Switching to an older version of DhcpEnumSubnetClientsV4 is pretty trivial. Unfortunately nothing similar to DhcpV4GetFreeIPAddress (which I use for free ip address lookups) is available in earlier versions of Windows -- I'll either need to drop the functionality for Windows server < 2012 or find another way of doing the lookup.

dmitri-d commented:
Rebased (no other changes).

dmitri-d commented:
[test]

dmitri-d commented:
@pikelly, @alexjfisher: this should now load and work on windows 2008. The only functionality that will be missing on 2008 is unused_ip -- the function isn't available prior to Windows 2012.

dmitri-d commented:
@pikelly, @alexjfisher: any chance you could give this a try?

pikelly commented:
@witlessbird: It is great that you now support W2008! I will check it now but my holiday season is still not quite done so I may be a few days.

alexjfisher commented:
@witlessbird Finally found a moment to try this out.  First thing I tried was a 'rebuild config' on a host.

```
D, [2016-09-28T16:58:53.593990 #112280] DEBUG -- : Loading subnets for \\10.60.248.70
D, [2016-09-28T16:58:53.593990 #112280] DEBUG -- : executing: c:\windows\sysnative\cmd.exe /c c:\Windows\System32\netsh.exe -c dhcp server \\10.60.248.70 show scope
I, [2016-09-28T16:58:54.061990 #112280]  INFO -- : Enumerated the scopes on \\10.60.248.70
E, [2016-09-28T16:58:54.061990 #112280] ERROR -- : undefined method `each_line' for #<Array:0x3645fd8>
D, [2016-09-28T16:58:54.061990 #112280] DEBUG -- : C:/foreman-smartproxy/testing/DHCP/smart-proxy/modules/dhcp/providers/server/native_ms.rb:184:in `find_all_subnets'
C:/foreman-smartproxy/testing/DHCP/smart-proxy/modules/dhcp/providers/server/native_ms.rb:197:in `loadSubnets'
```

alexjfisher commented:
@witlessbird Ignore that.  It'd help if I remembered to checkout the right branch!!!

alexjfisher commented:
@witlessbird I'm probably still doing something dumb, but at least I'm using the right branch now!

```
ERROR -- : Disabling all modules in the group ['dhcp_native_ms', 'dhcp'] due to a failure in one of them: cannot load such file -- dhcpsapi
```

I've done a bundle update to check the dhcpsapi gem is installed - it is.  What am I missing?

dmitri-d commented:
Try executing `bundler exec ruby bin/smart-proxy`?

alexjfisher commented:
I don't think I normally do that, do I?  I certainly haven't been whilst working on that DNS module change.
Anyway...
It looked like that might have worked.  It seem to load the module at least.  but then...

```
Segmentation fault
```

???!!

alexjfisher commented:
The seg fault isn't instantaneous.  If I'm quick, I've been able to do an import subnets (apparently successfully) before it dies.

dmitri-d commented:
Are you saying segfault happens even when you don't make any calls to smart-proxy? I'm away this week, any chance I could get access to your environment next week?

alexjfisher commented:
I'm not trying to make calls to the smart-proxy myself, but I assume foreman periodically is?
If i start the proxy using just `bin/smart-proxy`, (so that the module doesn't initialise), I can see the odd `GET /dhcp/...` come in and 404s being returned.  I think these are triggering the segfaults when I start up with `bundle exec`

I can probably arrange some sort of screen-share one day next week.  Meanwhile... enjoy the rest of your holiday!

dmitri-d commented:
Cheers. Is it the same url that gets called (when 404 is returned)? Could you post the log here?

dmitri-d commented:
Oh, also, are you running smart-proxy on windows 2008 server?

alexjfisher commented:
Yes, the proxy is on Windows 2008 R2.  I think the actual DHCP servers are on a mixture of 2008 and 2012.
I can probably get a new machine provisioned to run the proxy, but will have to leave the DHCP servers where they are.

alexjfisher commented:
> Cheers. Is it the same url that gets called (when 404 is returned)? Could you post the log here?

I'm assuming it's the same URL.  Nothing gets logged when it segfaults.
Here are  some example URLs/log messages that did come through after starting the proxy in such a way that the dhcp module doesn't get loaded.  Not sure why foreman has made these requests to the proxy.  Perhaps I had the proxy status page open somewhere?

```
foreman.example.com - - [28/Sep/2016:18:43:05 GMT Daylight Time] "GET /dhcp/10.61.53.0/00:50:56:ad:29:35 HTTP/1.1" 404 659
- -> /dhcp/10.61.53.0/00:50:56:ad:29:35
foreman.example.com - - [28/Sep/2016:18:43:05 GMT Daylight Time] "GET /dhcp/10.61.53.0/10.61.53.30 HTTP/1.1" 404 654
- -> /dhcp/10.61.53.0/10.61.53.30
```

dmitri-d commented:
@alexjfisher: I found and fixed an issue unrelated to the core dump you are seeing, but cannot make smart-proxy to core-dump in my environment (two windows 2012 servers with smart-proxy and dhcp server running on separate hosts). 

Any chance you could test these changes against windows 2012? Any chance I could use one of your windows 2008 environments to try to figure out the crash you are seeing?

pikelly commented:
Hi Witlessbird. I am also experiencing some problems. I have finally got W2K12R2 box and installed the dhcpsapi gem. I then  run this code and get a 1780 exception,  (null pointer).

@test_subnet = '192.168.1.0'
@test_ip     = '192.168.1.22'
@test_mac    = '01:01:02:03:04:06'
@first       = '192.168.1.31'
@second      = '192.168.1.32'

api = DhcpsApi::Server.new(server_name)

print_clients api

api.create_reservation(@test_ip, '255.255.255.0', @test_mac, 'solo')
api.set_reserved_option_value(66, @test_ip, @test_subnet, DhcpsApi::DHCP_OPTION_DATA_TYPE::DhcpStringDataOption,[@first])
sleep 2

print_clients api
pp api.get_client_by_ip_address @test_ip

the get_client_by_ip_address is segv'ing

dmitri-d commented:
@pikelly: I can't reproduce the segfault. I tried gem calling local windows2012 and remote windows 2012 (from another windows 2012 in the latter case). I used ruby 2.0 and 2.1.

Can you describe your environment please?  Are there any other reservations or leases in that subnet? Any other details you can think of?

dmitri-d commented:
Also fixed rubocop warnings.

pikelly commented:
The gem is running under ruby193 so maybe that is the reason for this problem. I was eventually hoping to run this under ruby187 but maybe I am being unrealistic.

dmitri-d commented:
Smart-proxy on windows doesn't support 1.8.7 (hasn't for a while now). Any chance you could test this on ruby 2+?

dmitri-d commented:
I looked into using ruby-dhcpsapi gem with ruby 1.9.3 and indeed it core dumps on a few calls (looks like when function uses pointer to a pointer to return results). I'm not as to the exact cause (GC or a bug in function binding code when pointer is a parameter), but I'm reluctant to spend any more time on this, as support for ruby 1.9.3 is going to be dropped after 1.14.

@alexjfisher, @pikelly: any chance you could test these changes with ruby version above 1.9.3?

pikelly commented:
@witlessbird : I tried the code on a W2K8 machine running under ruby200 and got this

pkelly@BRS____ ~
$ echo $OSVersion
W2K8R2

pkelly@BRS___ ~
$ ruby dhcptest.rb
C:/Ruby200/lib/ruby/gems/2.0.0/gems/dhcpsapi-0.0.5/lib/dhcpsapi/client.rb:143:in `get_client': Error retrieving client '192.168.1.22'. Unknown error '1780' (DhcpsApi::Error)
        from C:/Ruby200/lib/ruby/gems/2.0.0/gems/dhcpsapi-0.0.5/lib/dhcpsapi/client.rb:96:in`get_client_by_ip_address'
        from dhcptest.rb:31:in `<main>'

Sigh....

dmitri-d commented:
I found windows server 2008 R2 on MS Technet; will install it and see if I can replicate the issue.

alexjfisher commented:
On a Windows 2008 R2 server, I get the following segfault when running the dhcpsapi tests.

```
$ bundle exec rake
C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/server.rb:44: warning: assigned but unused variable - e
Run options: --seed 28786

# Running tests:

..EEEEE.C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/option_value.rb:378: [BUG] Segmentation fault
ruby 2.1.8p440 (2015-12-16 revision 53160) [i386-mingw32]

-- Control frame information -----------------------------------------------
c:0021 p:---- s:0118 e:000117 CFUNC  :DhcpGetOptionValueV5
c:0020 p:0096 s:0108 e:000107 METHOD C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/option_value.rb:378
c:0019 p:0029 s:0096 e:000095 METHOD C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/option_value.rb:145
c:0018 p:0287 s:0089 e:000086 METHOD C:/Users/afisher-admin/ruby-dhcpsapi/integration_test/option_value_test.rb:91
c:0017 p:0040 s:0079 e:000078 METHOD C:/Ruby21/lib/ruby/2.1.0/test/unit.rb:869
c:0016 p:0069 s:0074 e:000073 METHOD C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:1265
c:0015 p:0019 s:0066 e:000065 METHOD C:/Ruby21/lib/ruby/2.1.0/test/unit/testcase.rb:17
c:0014 p:0070 s:0062 e:000061 BLOCK  C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:940 [FINISH]
c:0013 p:---- s:0056 e:000055 CFUNC  :map
c:0012 p:0107 s:0053 e:000052 METHOD C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:933
c:0011 p:0011 s:0043 e:000042 BLOCK  C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:914 [FINISH]
c:0010 p:---- s:0040 e:000039 CFUNC  :map
c:0009 p:0008 s:0037 e:000036 METHOD C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:914
c:0008 p:0137 s:0032 e:000031 METHOD C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:884
c:0007 p:0009 s:0022 e:000021 METHOD C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:1092
c:0006 p:0009 s:0019 e:000018 BLOCK  C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:1079 [FINISH]
c:0005 p:---- s:0016 e:000015 CFUNC  :each
c:0004 p:0048 s:0013 e:000012 METHOD C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:1078
c:0003 p:0017 s:0009 e:000008 METHOD C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:1066
c:0002 p:0061 s:0005 E:0004d0 BLOCK  C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:802 [FINISH]
c:0001 p:0000 s:0002 E:0000f4 TOP    [FINISH]

C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:802:in `block in autorun'
C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:1066:in `run'
C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:1078:in `_run'
C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:1078:in `each'
C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:1079:in `block in _run'
C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:1092:in `run_tests'
C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:884:in `_run_anything'
C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:914:in `_run_suites'
C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:914:in `map'
C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:914:in `block in _run_suites'
C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:933:in `_run_suite'
C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:933:in `map'
C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:940:in `block in _run_suite'
C:/Ruby21/lib/ruby/2.1.0/test/unit/testcase.rb:17:in `run'
C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb:1265:in `run'
C:/Ruby21/lib/ruby/2.1.0/test/unit.rb:869:in `run_test'
C:/Users/afisher-admin/ruby-dhcpsapi/integration_test/option_value_test.rb:91:in `test_create_list_delete_reserved_option_value'
C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/option_value.rb:145:in `get_reserved_option_value'
C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/option_value.rb:378:in `dhcp_get_option_value_v5'
C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/option_value.rb:378:in `DhcpGetOptionValueV5'

-- C level backtrace information -------------------------------------------
C:\WINDOWS\SysWOW64\ntdll.dll(NtWaitForSingleObject+0x15) [0x77C5F8E1]
C:\WINDOWS\syswow64\kernel32.dll(WaitForSingleObjectEx+0x43) [0x77521194]
C:\WINDOWS\syswow64\kernel32.dll(WaitForSingleObject+0x12) [0x77521148]
C:\Ruby21\bin\msvcrt-ruby210.dll(rb_vm_bugreport+0xa7) [0x6D3893B7]
C:\Ruby21\bin\msvcrt-ruby210.dll(rb_name_err_mesg_new+0x69f) [0x6D2445AF]
C:\Ruby21\bin\msvcrt-ruby210.dll(rb_bug+0x2e) [0x6D24535E]
C:\Ruby21\bin\msvcrt-ruby210.dll(rb_check_safe_str+0x33b) [0x6D30B77B]
 [0x00401866]
C:\WINDOWS\SysWOW64\ntdll.dll(RtlKnownExceptionFilter+0xb7) [0x77CB5B67]

-- Other runtime information -----------------------------------------------

* Loaded script: C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/rake_test_loader.rb: OptionValueTest#test_create_list_delete_reserved_option_value

* Loaded features:

    0 enumerator.so
    1 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/enc/encdb.so
    2 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/enc/iso_8859_1.so
    3 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/enc/trans/transdb.so
    4 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/rbconfig.rb
    5 C:/Ruby21/lib/ruby/2.1.0/rubygems/compatibility.rb
    6 C:/Ruby21/lib/ruby/2.1.0/rubygems/defaults.rb
    7 C:/Ruby21/lib/ruby/2.1.0/rubygems/deprecate.rb
    8 C:/Ruby21/lib/ruby/2.1.0/rubygems/errors.rb
    9 C:/Ruby21/lib/ruby/2.1.0/rubygems/version.rb
   10 C:/Ruby21/lib/ruby/2.1.0/rubygems/requirement.rb
   11 C:/Ruby21/lib/ruby/2.1.0/rubygems/platform.rb
   12 C:/Ruby21/lib/ruby/2.1.0/rubygems/basic_specification.rb
   13 C:/Ruby21/lib/ruby/2.1.0/rubygems/stub_specification.rb
   14 C:/Ruby21/lib/ruby/2.1.0/rubygems/util/stringio.rb
   15 C:/Ruby21/lib/ruby/2.1.0/rubygems/specification.rb
   16 C:/Ruby21/lib/ruby/2.1.0/rubygems/exceptions.rb
   17 C:/Ruby21/lib/ruby/2.1.0/rubygems/defaults/operating_system.rb
   18 C:/Ruby21/lib/ruby/2.1.0/rubygems/core_ext/kernel_gem.rb
   19 thread.rb
   20 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/thread.so
   21 C:/Ruby21/lib/ruby/2.1.0/monitor.rb
   22 C:/Ruby21/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb
   23 C:/Ruby21/lib/ruby/2.1.0/rubygems.rb
   24 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/pathname.so
   25 C:/Ruby21/lib/ruby/2.1.0/pathname.rb
   26 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/constants.rb
   27 C:/Ruby21/lib/ruby/2.1.0/rubygems/dependency.rb
   28 C:/Ruby21/lib/ruby/2.1.0/rubygems/path_support.rb
   29 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/io/console.so
   30 C:/Ruby21/lib/ruby/2.1.0/rubygems/user_interaction.rb
   31 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/etc.so
   32 C:/Ruby21/lib/ruby/2.1.0/rubygems/config_file.rb
   33 C:/Ruby21/lib/ruby/2.1.0/rubygems/ext/build_error.rb
   34 C:/Ruby21/lib/ruby/2.1.0/rubygems/ext/builder.rb
   35 C:/Ruby21/lib/ruby/2.1.0/rubygems/ext/configure_builder.rb
   36 C:/Ruby21/lib/ruby/2.1.0/fileutils.rb
   37 C:/Ruby21/lib/ruby/2.1.0/delegate.rb
   38 C:/Ruby21/lib/ruby/2.1.0/tmpdir.rb
   39 C:/Ruby21/lib/ruby/2.1.0/tempfile.rb
   40 C:/Ruby21/lib/ruby/2.1.0/rubygems/ext/ext_conf_builder.rb
   41 C:/Ruby21/lib/ruby/2.1.0/rubygems/ext/rake_builder.rb
   42 C:/Ruby21/lib/ruby/2.1.0/optparse.rb
   43 C:/Ruby21/lib/ruby/2.1.0/rubygems/command.rb
   44 C:/Ruby21/lib/ruby/2.1.0/rubygems/ext/cmake_builder.rb
   45 C:/Ruby21/lib/ruby/2.1.0/rubygems/ext.rb
   46 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/rubygems_integration.rb
   47 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/current_ruby.rb
   48 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/shared_helpers.rb
   49 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/environment_preserver.rb
   50 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/socket.so
   51 C:/Ruby21/lib/ruby/2.1.0/socket.rb
   52 C:/Ruby21/lib/ruby/2.1.0/timeout.rb
   53 C:/Ruby21/lib/ruby/2.1.0/net/protocol.rb
   54 C:/Ruby21/lib/ruby/2.1.0/uri/common.rb
   55 C:/Ruby21/lib/ruby/2.1.0/uri/generic.rb
   56 C:/Ruby21/lib/ruby/2.1.0/uri/ftp.rb
   57 C:/Ruby21/lib/ruby/2.1.0/uri/http.rb
   58 C:/Ruby21/lib/ruby/2.1.0/uri/https.rb
   59 C:/Ruby21/lib/ruby/2.1.0/uri/ldap.rb
   60 C:/Ruby21/lib/ruby/2.1.0/uri/ldaps.rb
   61 C:/Ruby21/lib/ruby/2.1.0/uri/mailto.rb
   62 C:/Ruby21/lib/ruby/2.1.0/uri.rb
   63 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/zlib.so
   64 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/stringio.so
   65 C:/Ruby21/lib/ruby/2.1.0/net/http/exceptions.rb
   66 C:/Ruby21/lib/ruby/2.1.0/net/http/header.rb
   67 C:/Ruby21/lib/ruby/2.1.0/net/http/generic_request.rb
   68 C:/Ruby21/lib/ruby/2.1.0/net/http/request.rb
   69 C:/Ruby21/lib/ruby/2.1.0/net/http/requests.rb
   70 C:/Ruby21/lib/ruby/2.1.0/net/http/response.rb
   71 C:/Ruby21/lib/ruby/2.1.0/net/http/responses.rb
   72 C:/Ruby21/lib/ruby/2.1.0/net/http/proxy_delta.rb
   73 C:/Ruby21/lib/ruby/2.1.0/net/http/backward.rb
   74 C:/Ruby21/lib/ruby/2.1.0/net/http.rb
   75 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/date_core.so
   76 C:/Ruby21/lib/ruby/2.1.0/date/format.rb
   77 C:/Ruby21/lib/ruby/2.1.0/date.rb
   78 C:/Ruby21/lib/ruby/2.1.0/time.rb
   79 C:/Ruby21/lib/ruby/2.1.0/rubygems/request.rb
   80 C:/Ruby21/lib/ruby/2.1.0/cgi/core.rb
   81 C:/Ruby21/lib/ruby/2.1.0/cgi/util.rb
   82 C:/Ruby21/lib/ruby/2.1.0/cgi/cookie.rb
   83 C:/Ruby21/lib/ruby/2.1.0/cgi.rb
   84 C:/Ruby21/lib/ruby/2.1.0/rubygems/uri_formatter.rb
   85 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/fcntl.so
   86 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/digest.so
   87 C:/Ruby21/lib/ruby/2.1.0/digest.rb
   88 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/openssl.so
   89 C:/Ruby21/lib/ruby/2.1.0/openssl/bn.rb
   90 C:/Ruby21/lib/ruby/2.1.0/openssl/cipher.rb
   91 C:/Ruby21/lib/ruby/2.1.0/openssl/config.rb
   92 C:/Ruby21/lib/ruby/2.1.0/openssl/digest.rb
   93 C:/Ruby21/lib/ruby/2.1.0/openssl/x509.rb
   94 C:/Ruby21/lib/ruby/2.1.0/openssl/buffering.rb
   95 C:/Ruby21/lib/ruby/2.1.0/openssl/ssl.rb
   96 C:/Ruby21/lib/ruby/2.1.0/openssl.rb
   97 C:/Ruby21/lib/ruby/2.1.0/securerandom.rb
   98 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/fiddle.so
   99 C:/Ruby21/lib/ruby/2.1.0/fiddle/function.rb
  100 C:/Ruby21/lib/ruby/2.1.0/fiddle/closure.rb
  101 C:/Ruby21/lib/ruby/2.1.0/fiddle.rb
  102 C:/Ruby21/lib/ruby/2.1.0/fiddle/value.rb
  103 C:/Ruby21/lib/ruby/2.1.0/fiddle/pack.rb
  104 C:/Ruby21/lib/ruby/2.1.0/fiddle/struct.rb
  105 C:/Ruby21/lib/ruby/2.1.0/fiddle/cparser.rb
  106 C:/Ruby21/lib/ruby/2.1.0/fiddle/import.rb
  107 C:/Ruby21/lib/ruby/2.1.0/win32/importer.rb
  108 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/enc/utf_16le.so
  109 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/enc/trans/utf_16_32.so
  110 C:/Ruby21/lib/ruby/2.1.0/win32/registry.rb
  111 C:/Ruby21/lib/ruby/2.1.0/win32/resolv.rb
  112 C:/Ruby21/lib/ruby/2.1.0/resolv.rb
  113 C:/Ruby21/lib/ruby/2.1.0/rubygems/remote_fetcher.rb
  114 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/gem_remote_fetcher.rb
  115 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/gem_helpers.rb
  116 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/match_platform.rb
  117 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/rubygems_ext.rb
  118 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/version.rb
  119 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/errors.rb
  120 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler.rb
  121 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/settings.rb
  122 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/strscan.so
  123 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/source.rb
  124 C:/Ruby21/lib/ruby/2.1.0/i386-mingw32/digest/sha1.so
  125 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/source/path.rb
  126 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/source/git.rb
  127 C:/Ruby21/lib/ruby/2.1.0/rubygems/text.rb
  128 C:/Ruby21/lib/ruby/2.1.0/rubygems/name_tuple.rb
  129 C:/Ruby21/lib/ruby/2.1.0/rubygems/spec_fetcher.rb
  130 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/source/rubygems.rb
  131 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/lockfile_parser.rb
  132 C:/Ruby21/lib/ruby/2.1.0/set.rb
  133 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/definition.rb
  134 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/dependency.rb
  135 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/ruby_dsl.rb
  136 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/dsl.rb
  137 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/source_list.rb
  138 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/version.rb
  139 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/lazy_specification.rb
  140 C:/Ruby21/lib/ruby/2.1.0/tsort.rb
  141 C:/Ruby21/lib/ruby/2.1.0/forwardable.rb
  142 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/spec_set.rb
  143 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/index.rb
  144 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/ui.rb
  145 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/ui/silent.rb
  146 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/ui/rg_proxy.rb
  147 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/environment.rb
  148 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/runtime.rb
  149 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/remote_specification.rb
  150 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/dep_proxy.rb
  151 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendor/molinillo/lib/molinillo/gem_metadata.rb
  152 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendor/molinillo/lib/molinillo/errors.rb
  153 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendor/molinillo/lib/molinillo/dependency_graph.rb
  154 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendor/molinillo/lib/molinillo/state.rb
  155 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendor/molinillo/lib/molinillo/modules/specification_provider.rb
  156 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendor/molinillo/lib/molinillo/resolution.rb
  157 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendor/molinillo/lib/molinillo/resolver.rb
  158 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendor/molinillo/lib/molinillo/modules/ui.rb
  159 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendor/molinillo/lib/molinillo.rb
  160 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendored_molinillo.rb
  161 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/resolver.rb
  162 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/endpoint_specification.rb
  163 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/stub_specification.rb
  164 C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/setup.rb
  165 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/version.rb
  166 C:/Ruby21/lib/ruby/2.1.0/singleton.rb
  167 C:/Ruby21/lib/ruby/2.1.0/ostruct.rb
  168 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/ext/core.rb
  169 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/ext/string.rb
  170 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/win32.rb
  171 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/linked_list.rb
  172 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/cpu_counter.rb
  173 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/scope.rb
  174 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/task_argument_error.rb
  175 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/rule_recursion_overflow_error.rb
  176 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/task_manager.rb
  177 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/cloneable.rb
  178 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/file_utils.rb
  179 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/file_utils_ext.rb
  180 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/file_list.rb
  181 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/promise.rb
  182 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/thread_pool.rb
  183 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/private_reader.rb
  184 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/thread_history_display.rb
  185 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/trace_output.rb
  186 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/application.rb
  187 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/rake_module.rb
  188 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/pseudo_status.rb
  189 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/task_arguments.rb
  190 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/invocation_chain.rb
  191 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/invocation_exception_mixin.rb
  192 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/task.rb
  193 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/early_time.rb
  194 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/file_task.rb
  195 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/file_creation_task.rb
  196 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/multi_task.rb
  197 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/dsl_definition.rb
  198 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/default_loader.rb
  199 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/late_time.rb
  200 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/name_space.rb
  201 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/backtrace.rb
  202 C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake.rb
  203 C:/Ruby21/lib/ruby/2.1.0/minitest/unit.rb
  204 C:/Ruby21/lib/ruby/2.1.0/minitest/spec.rb
  205 C:/Ruby21/lib/ruby/2.1.0/minitest/mock.rb
  206 C:/Ruby21/lib/ruby/2.1.0/minitest/autorun.rb
  207 C:/Ruby21/lib/ruby/2.1.0/prettyprint.rb
  208 C:/Ruby21/lib/ruby/2.1.0/pp.rb
  209 C:/Ruby21/lib/ruby/2.1.0/test/unit/assertions.rb
  210 C:/Ruby21/lib/ruby/2.1.0/test/unit/testcase.rb
  211 C:/Ruby21/lib/ruby/2.1.0/test/unit.rb
  212 C:/Users/afisher-admin/ruby-dhcpsapi/integration_test/test_helper.rb
  213 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/2.1/ffi_c.so
  214 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/platform.rb
  215 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/types.rb
  216 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/library.rb
  217 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/errno.rb
  218 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/pointer.rb
  219 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/memorypointer.rb
  220 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/struct_layout_builder.rb
  221 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/struct.rb
  222 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/union.rb
  223 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/managedstruct.rb
  224 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/callback.rb
  225 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/io.rb
  226 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/autopointer.rb
  227 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/variadic.rb
  228 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/enum.rb
  229 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi/ffi.rb
  230 C:/Ruby21/lib/ruby/gems/2.1.0/gems/ffi-1.9.14-x86-mingw32/lib/ffi.rb
  231 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/enums.rb
  232 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/common.rb
  233 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/data_structures.rb
  234 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/error.rb
  235 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/class.rb
  236 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/client.rb
  237 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/misc.rb
  238 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/option.rb
  239 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/option_value.rb
  240 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/reservation.rb
  241 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/subnet.rb
  242 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/win2008/free_memory.rb
  243 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/win2008/class.rb
  244 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/win2008/client.rb
  245 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/win2008/option.rb
  246 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/win2008/option_value.rb
  247 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/win2008/subnet_element.rb
  248 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/win2008/subnet.rb
  249 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi/server.rb
  250 C:/Users/afisher-admin/ruby-dhcpsapi/lib/dhcpsapi.rb
  251 C:/Users/afisher-admin/ruby-dhcpsapi/integration_test/class_test.rb
  252 C:/Users/afisher-admin/ruby-dhcpsapi/integration_test/client_test.rb
  253 C:/Users/afisher-admin/ruby-dhcpsapi/integration_test/misc_test.rb
  254 C:/Users/afisher-admin/ruby-dhcpsapi/integration_test/option_test.rb
  255 C:/Users/afisher-admin/ruby-dhcpsapi/integration_test/option_value_test.rb
  256 C:/Users/afisher-admin/ruby-dhcpsapi/integration_test/reservation_test.rb
  257 C:/Users/afisher-admin/ruby-dhcpsapi/integration_test/subnet_test.rb

[NOTE]
You may have encountered a bug in the Ruby interpreter or extension libraries.
Bug reports are welcome.
For details: http://www.ruby-lang.org/bugreport.html


This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
rake aborted!
Command failed with status (3): [ruby -w -I"C:/Users/afisher-admin/ruby-dhcpsapi/integration_test;lib" -I"C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib" "C:/Ruby21/lib/ruby/gems/2.1.0/gems/rake-11.3.0/lib/rake/rake_test_loader.rb" "integration_test/class_test.rb" "integration_test/client_test.rb" "integration_test/misc_test.rb" "integration_test/option_test.rb" "integration_test/option_value_test.rb" "integration_test/reservation_test.rb" "integration_test/subnet_test.rb" ]

Tasks: TOP => default => test
(See full trace by running task with --trace)
```

dmitri-d commented:
@pikelly, @alexjfisher: I fixed the issues that were causing coredumps on Windows 2008: in a few places Windows api documentation specifies passing function parameters by value, while in fact pointers are expected in those spots (no idea why it worked on Windows2012). I released a new version of ruby-dhcpsapi gem with the changes.

dmitri-d commented:
@alexjfisher: Fixed a couple of issues you discovered: 
- reservations are now being created with "DHCP" client type (this requires dhcpsapi gem update).
- 'hostname' dhcp option is now being set.  'PXEClient' option will be reset to an empty string if the option is defined for the given scope.

I couldn't reproduce the issue with reservations not appearing. Smart-proxy api always returns the data, but there was a bug related to 'hostname' dhcp option not being set.

dmitri-d commented:
@alexjfisher: Resolved issues with mis-identification of leases and reservations. This required a binding for an additional native dhcps call, so you'll need to run bundle update.

alexjfisher commented:
@witlessbird Just retested.  Looking good.

Running the proxy on Win 2008 I've tested the following against a win2012 DHCP server...
- Listing and accessing reservation details with the [DHCP-browser plugin](https://github.com/theforeman/foreman_dhcp_browser)
- 'Rebuild Config' on a host
- Creating a new host
- Deleting a host

Also tested against a Win 2008 DHCP server
- DHCP browsing
- 'Rebuild Config'

dmitri-d commented:
@alexjfisher, sounds great. Thanks for testing these changes! Also, [test].

dmitri-d commented:
smart-proxy-win-service now uses bundler to load dependencies. This was required as windows-specific runtime dependencies are in a non-default bundler group.

domcleal commented:
> Move ticket #7766 to the Foreman project

Sorry, this was a bug in this new warning feature. It's correct about the ticket being in the wrong project, but it was meant to say that it should be in the _Smart Proxy_ project.

dmitri-d commented:
> Sorry, this was a bug in this new warning feature. It's correct about the ticket being in the wrong project, but it was meant to say that it should be in the Smart Proxy project.

Just moved it to Smart-Proxy.

domcleal commented on bin/smart-proxy-win-service:
```diff
@@ -2,41 +2,42 @@ $LOAD_PATH.unshift(*Dir[File.expand_path("../../lib", __FILE__), File.expand_pat
 Dir.mkdir(LOG_DIR_PATH) unless Dir.exist?(LOG_DIR_PATH)
 
-def logger
-  @logger ||= Logger.new(File.expand_path("win-service.log", LOG_DIR_PATH))
```
Why remove Logger here? It has timestamps etc by default that are probably useful.

dmitri-d replied:
Opening a file directly is more reliable (it's amazing how much grief these few lines of code give me).

domcleal commented on bundler.d/windows.rb:
```diff
@@ -1,4 +1,5 @@
 group :windows do
   gem 'highline', :platforms => [:mingw, :x64_mingw]
   gem 'win32-service', :platforms => [:mingw, :x64_mingw]
+  gem 'dhcpsapi'
```
Should this be pinned defensively? e.g. `~>0.0` so API changes are easier.

domcleal commented on modules/dhcp_common/subnet.rb:
```diff
@@ -22,14 +22,12 @@ def initialize network, netmask, options = {}
       @netmask = validate_ip netmask
       @options = {}
 
-      @options[:routers] = options[:routers].each{|ip| validate_ip ip } if options[:routers]
```
Why remove the validations from this class?

domcleal commented on modules/dhcp_native_ms/dhcp_native_ms_main.rb:
```diff
@@ -3,279 +3,228 @@
+                                                    .map {|r| r[:element][:reserved_ip_address]})
+      if reservation_subnet_elements_ips.include?(client[:client_ip_address])
+        standard_option_values = standard_option_values(dhcpsapi.list_reserved_option_values(client[:client_ip_address], subnet_address))
+        #vendor_option_values = vendor_option_values(dhcpsapi.list_reserved_option_values(client[:client_ip_address], subnet_address, vendor), vendor)
```
Why's this commented (ditto in a few places below)?

domcleal commented on modules/dhcp_native_ms/dhcp_native_ms_main.rb:
```diff
@@ -3,279 +3,228 @@
+              :hostname => client[:client_name],
+              :deleteable => true}
+      opts.merge!(options)
+      logger.debug opts.inspect
```
This logging needs more context in the message or to be removed.

domcleal commented on modules/dhcp_native_ms/dhcp_native_ms_main.rb:
```diff
@@ -3,279 +3,228 @@
+              :hostname => (client[:client_name] || '*lease*'),
+              :ends => client[:client_lease_expires]}
+      opts.merge!(options)
+      logger.debug opts.inspect
```
ditto

domcleal commented on modules/dhcp_native_ms/dhcp_native_ms_main.rb:
```diff
@@ -3,279 +3,228 @@
+    end
+
+    def dhcps_option_type_from_sunw_kind(kind)
+      return case kind
```
`return` is unnecessary here, easier to read with just the `case`

dmitri-d commented:
- Added timestamps for windows service logging. 
- pinned dhcpsapi gem
- Restored validations in Subnet constructor
- Removed all invocations of vendor options-related code. 
  - It's broken in a couple of places
  - Doesn't appear anyone is using sun-specific vendor options in MS environment (does anyone anymore anywhere?)
- Logging of validation errors during reservation/lease object creation should be clearer now

dmitri-d commented:
@alexjfisher: `Proxy::DHCP::NativeMS::Provider#find_record` now returns `nil` instead of throwing an exception when record doesn't exist.

dmitri-d commented:
Client mac addresses now use lowercase letters, as mac address comparison is case-sensitive on foreman side.

dmitri-d commented:
Removed netsh reference from user notification in windows service registration script.

dmitri-d commented:
[test]

domcleal commented on test/dhcp/subnet_test.rb:
```diff
@@ -19,39 +19,6 @@ def test_should_not_save_invalid_network_addresses
     end
   end
 
-  def test_should_not_save_invalid_router_addresses
```
Could you restore these tests please, to go with the restored code in Subnet?

domcleal commented:
@alexjfisher if you've been testing this, are you satisfied with the current state? I won't be testing this PR, so I'll take your word for it.

alexjfisher commented:
@domcleal I've been running it in production on a couple of subnets since early last week.  Since the last couple of fixes, I've not found any more issues. Thanks @witlessbird !

That said, I wouldn't be too surprised if there were still a couple of corner cases out there that someone else might hit - perhaps running a different Windows version or patch level or something??

Maybe you could get some extra volunteer testers from the mailing list or IRC?  The performance improvement should make it an easy sell!  For me, 'rebuild config' on a single host went from almost 2 minutes to 2-3 seconds.

dmitri-d commented:
restored subnet tests. 

> Maybe you could get some extra volunteer testers from the mailing list or IRC?

I tried looking for volunteers to help with testing but other than @alexjfisher and @pikelly offered help.

dmitri-d commented:
'hostname' attribute is now always populated: either by wins/netbios name, or 'hostname' dhcp option if it  is available.

dmitri-d commented:
rebased.

dmitri-d commented:
Removed a reference to 'netsh' from the example config file.

alexjfisher commented:
The only other thing that I've noticed is https://github.com/theforeman/theforeman.org/blob/gh-pages/_includes/manuals/1.13/4.3.1_smartproxy_installation.md#test-and-configure-smart-proxy-features will need to be updated.

> Open an administrative command prompt and run `ruby <smart-proxy location>\bin\smart-proxy`

now needs `bundle exec`.

The 1.14 docs aren't branched yet, so I can't create a PR.  But after this is merged, perhaps someone can add a task to https://github.com/theforeman/theforeman.org/issues/708

BTW, the demo I gave today, was on a new Windows 2012 server.  All my previous testing had been on 2008 R2.

domcleal commented:
Merged, thanks very much @witlessbird and @alexjfisher.

(Note added to docs TODO list.)
